context_header.go

```diff
@@ -38,29 +38,65 @@ type ContextWithHeaders interface {
 
 type headerCtx struct {
 	context.Context
+}
+
+// headersContainer stores the headers, and is itself stored in the context under `contextKeyHeaders`
+type headersContainer struct {
 	reqHeaders  map[string]string
 	respHeaders map[string]string
 }
 
+func (c *headerCtx) headers() *headersContainer {
+	if h, ok := c.Value(contextKeyHeaders).(*headersContainer); ok {
+		return h
+	}
+	return nil
+}
+
 // Headers gets application headers out of the context.
 func (c *headerCtx) Headers() map[string]string {
-	return c.reqHeaders
+	if h := c.headers(); h != nil {
+		return h.reqHeaders
+	}
+	return nil
 }
 
 // ResponseHeaders returns the response headers.
 func (c *headerCtx) ResponseHeaders() map[string]string {
-	return c.respHeaders
+	if h := c.headers(); h != nil {
+		return h.respHeaders
+	}
+	return nil
 }
 
 // SetResponseHeaders sets the response headers.
 func (c *headerCtx) SetResponseHeaders(headers map[string]string) {
-	c.respHeaders = headers
+	if h := c.headers(); h != nil {
+		h.respHeaders = headers
+		return
+	}
+	panic("SetResponseHeaders called on ContextWithHeaders not created via WrapWithHeaders")
 }
 
 // WrapWithHeaders returns a Context that can be used to make a call with request headers.
+// If the parent `ctx` is already an instance of ContextWithHeaders, its Headers will be
+// carried over, but the new `headers` will take precedence. ResponseHeaders are not copied.
 func WrapWithHeaders(ctx context.Context, headers map[string]string) ContextWithHeaders {
-	return &headerCtx{
-		Context:    ctx,
+	h := &headersContainer{
 		reqHeaders: headers,
 	}
+	if parent_h, ok := ctx.Value(contextKeyHeaders).(*headersContainer); ok {
+		if pl := len(parent_h.reqHeaders); pl > 0 {
+			mergedHeaders := make(map[string]string, len(headers) + pl)
+			for k, v := range parent_h.reqHeaders {
+				mergedHeaders[k] = v
+			}
+			for k, v := range headers {
+				mergedHeaders[k] = v
+			}
+			h.reqHeaders = mergedHeaders
+		}
+	}
```

Comment:
Sorry for the back and forth, but I don't think this is the way we want to solve this issue since it becomes really hard to override all headers (e.g. you want to drop the existing ones and add new ones).

So we have a few use cases, and I have code for what I think each of those should look like:

 *  Service gets incoming call, wants to carry over headers, and add a new header
```go
ctx := NewContextBuilder(...)
  .SetParentContext(ctx)
  .AddHeader("extra", "value")
  .Build()
```

Headers are merged because of use of "AddHeader". If the user uses "SetHeaders" that should replace. Unless there's a "merge" in the name, I think previous headers should be replaced.

 * Service gets incoming call, and wants to propagate them all
```go
// ctx can be directly used
```

 * Service gets incoming call, wants to drop all existing headers, and set a new bunch of headers
```go
ctx = WrapWIthHeaders(ctx, map[string]string{"new": "headers"})
```


(we should have separate tests for these use cases and make sure it works fine)